### ISSUE TITLE
noteHistoryMetadataObservation(): CHECK constraint failed: moz_places_metadata

### ISSUE THREAD

**Issue by pocmo** [mozilla/application-services#4227]
Fenix issue:
https://github.com/mozilla-mobile/fenix/issues/20123

```
mozilla.appservices.places.PlacesException: Error executing SQL: CHECK constraint failed: moz_places_metadata
at mozilla.appservices.places.RustError.intoException(RustError.kt:58)
at mozilla.appservices.places.PlacesWriterConnection.noteHistoryMetadataObservation(PlacesConnection.kt:1865)
at mozilla.components.browser.storage.sync.PlacesHistoryStorage.noteHistoryMetadataObservation$suspendImpl(PlacesHistoryStorage.kt:263)
at mozilla.components.browser.storage.sync.PlacesHistoryStorage.noteHistoryMetadataObservation(Unknown Source:0)
at org.mozilla.fenix.historymetadata.DefaultHistoryMetadataService$createMetadata$1.invokeSuspend(HistoryMetadataService.kt:77)
```

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2668)
┆Epic: History Metadata Sync

**Comment by pocmo** [comment]
@grigoryk and @csadilek are both out this week.

**Comment by csadilek** [comment]
This should only be active in debug builds. See https://github.com/mozilla-mobile/fenix/issues/20123#issuecomment-866285339 for how to disable. We'll look into it when we're back :)

**Comment by rfk** [comment]
> data-sync-user closed this 3 minutes ago

Sorry, there are some Jira automation issues happening at the moment, this was not closed on purpose. I'm not going to touch it in case I make things worse, but hopefully this'll be re-opened shortly.

**Comment by csadilek** [comment]
TL;DR Closing this as we fixed it in A-C / Fenix.

There's only one CHECK constraint for history metadata, which was added to make sure that the referrer place ID and ID for which we are recording are not identical i.e., that we're not saving a metadata record which references itself as a referrer: https://github.com/mozilla/application-services/blob/01034805ed198220c4e26c8f70c70aefbb04e66b/components/places/sql/create_shared_schema.sql#L265

This check makes perfect sense to keep. However, we recently loosened our definition of referrer to not just cover parent tabs, but also the pages the user navigated from directly (in the same tab). There was a bug in Fenix though causing us to record referrers for all page navigations, including incorrect ones :).

We also made sure to handle `PlacesException` for history metadata in A-C and logging it to Sentry instead of causing a crash. Same as for regular history writes.
